File: ios/ZoltoPOS/ZoltoPOS/Services/OfflinePaymentManager.swift
import Foundation
import SwiftData
import Combine

/**
 * Central coordinator for offline payment support on iOS.
 *
 * - Records every sale locally (as [PendingTransactionModel]) as soon as it
 *   happens, regardless of connectivity.
 * - When online, attempts immediate sync; when offline, the transaction stays
 *   pending until connectivity returns and [syncAllPending] is called.
 * - Provides [pendingCount] via Combine so the UI can show a badge.
 */
@MainActor
class OfflinePaymentManager: ObservableObject {
    static let shared = OfflinePaymentManager()

    private let networkMonitor = NetworkMonitor()
    private var modelContext: ModelContext?
    private var cancellables = Set<AnyCancellable>()

    /// True when the device has an active internet connection.
    @Published var isOnline: Bool = true

    /// Number of transactions waiting to be synced.
    @Published var pendingCount: Int = 0

    private init() {
        networkMonitor.$isOnline
            .receive(on: DispatchQueue.main)
            .assign(to: &$isOnline)

        networkMonitor.start()

        // Poll pending count periodically
        Timer.publish(every: 2.0, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                Task { @MainActor in
                    await self?.refreshPendingCount()
                }
            }
            .store(in: &cancellables)
    }

    /// Must be called after the SwiftData ModelContext is available (e.g. from
    /// the App's body or the first View `.onAppear`).
    func configure(with context: ModelContext) {
        self.modelContext = context
        Task { @MainActor in
            await refreshPendingCount()
        }
    }

    // MARK: - Recording transactions

    /// Records a cash sale locally and immediately attempts to sync if online.
    func recordCashSale(
        productIds: [Int],
        allowHidden: Bool,
        priceOverrides: [String: Int],
        customItems: [CustomLineItemRequest],
        totalRappen: Int,
        itemCount: Int
    ) async throws -> UUID {
        let payload = PendingTransactionPayload(
            productIds: productIds,
            priceOverrides: priceOverrides,
            customItems: customItems.map { CustomItemPayload(name: $0.name, priceRappen: $0.priceRappen) },
            paymentMethod: "cash",
            allowHidden: allowHidden
        )
        let tx = PendingTransactionModel(
            transactionType: "cash",
            payloadJson: PendingTransactionSerializer.toJson(payload),
            totalRappen: totalRappen,
            displayLabel: "\(itemCount) items - Cash"
        )

        modelContext?.insert(tx)
        try? modelContext?.save()
        await refreshPendingCount()

        if networkMonitor.isCurrentlyOnline {
            _ = await syncTransaction(tx)
        }
        return tx.localId
    }

    /// Records a card-payment backend confirmation that failed after the
    /// card was already charged by Stripe. Queued for later sync.
    func recordCardBackendConfirm(
        paymentIntentId: String,
        posOrderId: Int,
        totalRappen: Int,
        itemCount: Int
    ) async -> UUID {
        let payload = PendingTransactionPayload(
            paymentMethod: "card",
            paymentIntentId: paymentIntentId,
            posOrderId: posOrderId
        )
        let tx = PendingTransactionModel(
            transactionType: "card_backend_confirm",
            payloadJson: PendingTransactionSerializer.toJson(payload),
            totalRappen: totalRappen,
            displayLabel: "\(itemCount) items - Card (confirm)"
        )

        modelContext?.insert(tx)
        try? modelContext?.save()
        await refreshPendingCount()

        if networkMonitor.isCurrentlyOnline {
            _ = await syncTransaction(tx)
        }
        return tx.localId
    }

    /// Records a TWINT backend confirmation that failed after the customer
    /// already paid. Queued for later sync.
    func recordTwintBackendConfirm(
        paymentIntentId: String,
        posOrderId: Int,
        totalRappen: Int,
        itemCount: Int
    ) async -> UUID {
        let payload = PendingTransactionPayload(
            paymentMethod: "twint",
            paymentIntentId: paymentIntentId,
            posOrderId: posOrderId
        )
        let tx = PendingTransactionModel(
            transactionType: "twint",
            payloadJson: PendingTransactionSerializer.toJson(payload),
            totalRappen: totalRappen,
            displayLabel: "\(itemCount) items - TWINT (confirm)"
        )

        modelContext?.insert(tx)
        try? modelContext?.save()
        await refreshPendingCount()

        if networkMonitor.isCurrentlyOnline {
            _ = await syncTransaction(tx)
        }
        return tx.localId
    }

    // MARK: - Sync

    /// Attempts to sync a single pending transaction. Returns true on success.
    func syncTransaction(_ tx: PendingTransactionModel) async -> Bool {
        tx.status = "syncing"
        try? modelContext?.save()

        do {
            let payload = PendingTransactionSerializer.fromJson(tx.payloadJson)
            switch tx.transactionType {
            case "cash":
                let _ = try await ApiService.shared.manualSale(
                    productIds: payload.productIds,
                    paymentMethod: "cash",
                    // Replaying without this is why a queued sale of a hidden
                    // piece came back 409 "no longer available" forever.
                    allowHidden: payload.allowHidden,
                    priceOverrides: payload.priceOverrides,
                    customItems: payload.customItems.map {
                        CustomLineItemRequest(name: $0.name, priceRappen: $0.priceRappen)
                    }
                )
            case "card_backend_confirm":
                guard let piId = payload.paymentIntentId else {
                    throw NSError(domain: "OfflinePayment", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing paymentIntentId"])
                }
                let _ = try await ApiService.shared.confirmSale(
                    posOrderId: payload.posOrderId ?? 0,
                    paymentIntentId: piId
                )
            case "twint":
                guard let piId = payload.paymentIntentId else {
                    throw NSError(domain: "OfflinePayment", code: 2, userInfo: [NSLocalizedDescriptionKey: "Missing paymentIntentId"])
                }
                let _ = try await ApiService.shared.confirmSale(
                    posOrderId: payload.posOrderId ?? 0,
                    paymentIntentId: piId
                )
            default:
                throw NSError(domain: "OfflinePayment", code: 3, userInfo: [NSLocalizedDescriptionKey: "Unknown transaction type: \(tx.transactionType)"])
            }

            tx.status = "synced"
            try? modelContext?.save()
            await refreshPendingCount()
            return true
        } catch {
            tx.status = "failed"
            tx.lastError = error.localizedDescription
            tx.retryCount += 1
            try? modelContext?.save()
            await refreshPendingCount()
            return false
        }
    }

    /// Retries all pending/failed transactions. Returns the number synced.
    func syncAllPending() async -> Int {
        guard let context = modelContext else { return 0 }
        let descriptor = FetchDescriptor<PendingTransactionModel>(
            predicate: #Predicate { $0.status == "pending" || $0.status == "failed" }
        )
        guard let pending = try? context.fetch(descriptor) else { return 0 }
        if pending.isEmpty { return 0 }

        var successCount = 0
        for tx in pending {
            if await syncTransaction(tx) { successCount += 1 }
        }
        await cleanupSynced()
        await refreshPendingCount()
        return successCount
    }

    /// Removes all "synced" transactions from the local store.
    func cleanupSynced() async {
        guard let context = modelContext else { return }
        let descriptor = FetchDescriptor<PendingTransactionModel>(
            predicate: #Predicate { $0.status == "synced" }
        )
        guard let synced = try? context.fetch(descriptor) else { return }
        for tx in synced {
            context.delete(tx)
        }
        try? context.save()
    }

    // MARK: - Internal

    private func refreshPendingCount() async {
        guard let context = modelContext else { return }
        let descriptor = FetchDescriptor<PendingTransactionModel>(
            predicate: #Predicate { $0.status == "pending" || $0.status == "failed" }
        )
        let count = (try? context.fetchCount(descriptor)) ?? 0
        pendingCount = count
    }
}

File: android/app/src/main/java/ch/zolto/pos/data/OfflinePaymentManager.kt
package ch.zolto.pos.data

import android.content.Context
import androidx.work.*
import ch.zolto.pos.data.local.*
import ch.zolto.pos.data.models.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.concurrent.TimeUnit

/**
 * Central coordinator for offline payment support.
 *
 * - Records every sale locally (into [PendingTransactionEntity]) as soon as it
 *   happens, regardless of connectivity.
 * - When online, attempts immediate sync; when offline, the transaction stays
 *   pending until [SyncWorker] or a manual retry picks it up.
 * - Provides a reactive [pendingCountFlow] so the UI can show a badge.
 */
class OfflinePaymentManager(context: Context) {

    private val db = DatabaseClient.getInstance(context)
    private val dao = db.pendingTransactionDao()
    private val networkMonitor = NetworkMonitor(context)
    private val workManager = WorkManager.getInstance(context)

    /** True when the device currently has internet access. */
    val isOnline: Flow<Boolean> = networkMonitor.isOnline

    /** Emits the current count of unsynced pending transactions. */
    val pendingCountFlow: Flow<Int> = flow {
        while (true) {
            emit(dao.getPendingCount())
            kotlinx.coroutines.delay(2000)
        }
    }

    /**
     * Records a cash sale locally and immediately attempts to sync it if
     * online. Call this from [PaymentViewModel.createCashSale].
     */
    suspend fun recordCashSale(
        productIds: List<Int>,
        allowHidden: Boolean,
        priceOverrides: Map<String, Int>,
        customItems: List<CustomLineItemRequest>,
        totalRappen: Int,
        itemCount: Int,
    ): Long {
        val payload = PendingTransactionPayload(
            productIds = productIds,
            priceOverrides = priceOverrides,
            customItems = customItems.map { CustomItemPayload(it.name, it.priceRappen) },
            paymentMethod = "cash",
            allowHidden = allowHidden,
        )
        val tx = PendingTransactionEntity(
            transactionType = "cash",
            payloadJson = PendingTransactionSerializer.toJson(payload),
            createdAt = isoNow(),
            totalRappen = totalRappen,
            displayLabel = "$itemCount items - Cash",
        )
        val id = dao.insert(tx)

        // Attempt immediate sync if we're online
        if (networkMonitor.isCurrentlyOnline()) {
            syncTransaction(dao.getPending().firstOrNull { it.id == id } ?: return id)
        } else {
            scheduleSyncWorker()
        }
        return id
    }

    /**
     * Records a card-payment backend confirmation that failed (e.g., network
     * hiccup after Stripe already charged the card). The sale is treated as
     * completed locally; we just need to tell our backend about it later.
     */
    suspend fun recordCardBackendConfirm(
        paymentIntentId: String,
        posOrderId: Int,
        totalRappen: Int,
        itemCount: Int,
    ): Long {
        val payload = PendingTransactionPayload(
            paymentIntentId = paymentIntentId,
            posOrderId = posOrderId,
            paymentMethod = "card",
        )
        val tx = PendingTransactionEntity(
            transactionType = "card_backend_confirm",
            payloadJson = PendingTransactionSerializer.toJson(payload),
            createdAt = isoNow(),
            totalRappen = totalRappen,
            displayLabel = "$itemCount items - Card (confirm)",
        )
        val id = dao.insert(tx)

        if (networkMonitor.isCurrentlyOnline()) {
            syncTransaction(dao.getPending().firstOrNull { it.id == id } ?: return id)
        } else {
            scheduleSyncWorker()
        }
        return id
    }

    /**
     * Records a locally-completed TWINT sale that couldn't be confirmed with
     * the backend due to a network error after the customer already paid.
     */
    suspend fun recordTwintBackendConfirm(
        paymentIntentId: String,
        posOrderId: Int,
        totalRappen: Int,
        itemCount: Int,
    ): Long {
        val payload = PendingTransactionPayload(
            paymentIntentId = paymentIntentId,
            posOrderId = posOrderId,
            paymentMethod = "twint",
        )
        val tx = PendingTransactionEntity(
            transactionType = "twint",
            payloadJson = PendingTransactionSerializer.toJson(payload),
            createdAt = isoNow(),
            totalRappen = totalRappen,
            displayLabel = "$itemCount items - TWINT (confirm)",
        )
        val id = dao.insert(tx)

        if (networkMonitor.isCurrentlyOnline()) {
            syncTransaction(dao.getPending().firstOrNull { it.id == id } ?: return id)
        } else {
            scheduleSyncWorker()
        }
        return id
    }

    /**
     * Attempts to sync a single pending transaction. Returns true on success.
     */
    suspend fun syncTransaction(tx: PendingTransactionEntity): Boolean {
        dao.markStatus(tx.id, "syncing")
        return try {
            val payload = PendingTransactionSerializer.fromJson(tx.payloadJson)
            val api = RetrofitClient.apiService
            when (tx.transactionType) {
                "cash" -> {
                    api.manualSale(
                        ManualSaleRequest(
                            productIds = payload.productIds,
                            paymentMethod = "cash",
                            // Replaying without this is why a queued sale of a
                            // hidden piece came back 409 forever.
                            allowHidden = payload.allowHidden,
                            priceOverrides = payload.priceOverrides,
                            customItems = payload.customItems.map {
                                CustomLineItemRequest(it.name, it.priceRappen)
                            },
                        )
                    )
                }
                "card_backend_confirm" -> {
                    val piId = payload.paymentIntentId
                        ?: throw IllegalStateException("Missing paymentIntentId")
                    api.confirmSale(SaleRequest(paymentIntentId = piId))
                }
                "twint" -> {
                    val piId = payload.paymentIntentId
                        ?: throw IllegalStateException("Missing paymentIntentId")
                    api.confirmSale(SaleRequest(paymentIntentId = piId))
                }
                else -> throw IllegalStateException("Unknown transaction type: ${tx.transactionType}")
            }
            dao.markStatus(tx.id, "synced")
            true
        } catch (e: Exception) {
            dao.markStatus(tx.id, "failed", e.message)
            false
        }
    }

    /**
     * Retries all pending and failed transactions. Returns the number
     * successfully synced.
     */
    suspend fun syncAllPending(): Int {
        val pending = dao.getPending()
        if (pending.isEmpty()) return 0
        var successCount = 0
        for (tx in pending) {
            if (syncTransaction(tx)) successCount++
        }
        // Clean up synced rows opportunistically
        dao.deleteSynced()
        return successCount
    }

    /**
     * Enqueues a one-time [SyncWorker] that will retry pending transactions
     * when connectivity is available. Uses a network constraint so WorkManager
     * only runs it when online.
    */
    private fun scheduleSyncWorker() {
        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()

        val request = OneTimeWorkRequestBuilder<SyncWorker>()
            .setConstraints(constraints)
            .setBackoffCriteria(
                BackoffPolicy.EXPONENTIAL,
                WorkRequest.MIN_BACKOFF_MILLIS,
                TimeUnit.MILLISECONDS
            )
            .addTag(SyncWorker.TAG)
            .build()

        workManager.enqueueUniqueWork(
            SyncWorker.NAME,
            ExistingWorkPolicy.KEEP,
            request
        )
    }

    private fun isoNow(): String = DateTimeFormatter.ISO_INSTANT.format(Instant.now())
}
